transmart-core-db/grails-app/services/org/transmartproject/db/userqueries/UserQueryService.groovy

```diff
@@ -1,46 +1,188 @@
+    }
+
+    @Override
+    UserQueryRepresentation create(UserQueryRepresentation representation, User currentUser) {
+        def query = new Query(username: currentUser.username)
+        if (representation.name ==  null || representation.name.trim().empty) {
+            throw new InvalidArgumentsException("Query name is empty.")
+        }
+        validateSubscriptionEnabled(representation.subscribed, representation.subscriptionFreq)
+        if (representation.subscribed) {
+            // Check query access when subscription is enabled
+            checkConstraintAccess(representation.patientsQuery, currentUser)
+        }
+
+        query.with {
+            name = representation.name
+            patientsQuery = representation.patientsQuery.toJson()
+            observationsQuery = BindingHelper.objectMapper.writeValueAsString(representation.observationsQuery)
+            bookmarked = representation.bookmarked ?: false
+            subscribed = representation.subscribed ?: false
+            subscriptionFreq = representation.subscriptionFreq
+            queryBlob = BindingHelper.objectMapper.writeValueAsString(representation.queryBlob)
+            apiVersion = representation.apiVersion
+        }
+
+        query = save(query, currentUser)
+
+        def result = toRepresentation(query)
+
+        if (query.subscribed) {
+            // Create initial patient set when subscription is enabled
+            userQuerySetResource.createSetWithInstances(result, currentUser)
+        }
+
+        result
     }
 
     @Override
-    void save(UserQuery query, User currentUser) {
+    UserQueryRepresentation update(Long id, UserQueryRepresentation representation, User currentUser) {
+        validateSubscriptionEnabled(representation.subscribed, representation.subscriptionFreq)
+
+        UserQuery query = fetch(id, currentUser)
+
+        if (representation.name != null) {
+            query.name = representation.name
+        }
+        if (representation.bookmarked != null) {
+            query.bookmarked = representation.bookmarked
+        }
+        boolean newSubscription = false
+        if (representation.subscribed != null) {
+            if (representation.subscribed) {
+                if (!query.patientsQuery) {
+                    throw new InvalidArgumentsException("Cannot subscribe to an empty query.")
+                }
+                // Check query access when subscription is enabled
+                checkConstraintAccess(ConstraintFactory.read(query.patientsQuery), currentUser)
+                if (!query.subscribed) {
+                    // This is a new subscription, an initial patient set needs to be generated
+                    newSubscription = true
+                }
+            }
+            query.subscribed = representation.subscribed
+        }
+        if (representation.subscriptionFreq != null) {
+            query.subscriptionFreq = representation.subscriptionFreq
```

Comment:
This field could be cleared when unsubscribing, but it is not that important, since it is not used anywhere, when not subscribed.